### PR TITLE
cache ItemNotFoundException in the CachedClient

### PR DESCRIPTION
**Pull request by lsmith77**
dropped the query count from 68 to 2

https://blackfire.io/profiles/compare/66c8148d-478c-4bd5-8864-986bc3a6bc66/graph

now the main overhead remains inside Jackalope core

### CODE REVIEW

File: src/Jackalope/Transport/DoctrineDBAL/CachedClient.php [hunk 1]
```diff
@@ -134,10 +134,23 @@ public function getNode($path)
 
         $cacheKey = "nodes: $path, ".$this->workspaceName;
         if (isset($this->caches['nodes']) && (false !== ($result = $this->caches['nodes']->fetch($cacheKey)))) {
+            if ('' === $result) {
+                throw new ItemNotFoundException("Item $path not found in workspace ".$this->workspaceName);
```
Comment (dantleech) [inline src/Jackalope/Transport/DoctrineDBAL/CachedClient.php]: Would be nice to use `sprintf` here and quote the values, `sprintf('Item "%s" not found in workspace "%s"' ..`

File: src/Jackalope/Transport/DoctrineDBAL/CachedClient.php [hunk 2]
```diff
@@ -134,10 +134,23 @@ public function getNode($path)
...
+
             return $result;
         }
 
-        $node = parent::getNode($path);
+        try {
+            $node = parent::getNode($path);
+        } catch (\Exception $e) {
+            if (isset($this->caches['nodes'])) {
+                $this->caches['nodes']->save($cacheKey, '');
```
Comment (dantleech) [inline src/Jackalope/Transport/DoctrineDBAL/CachedClient.php]: Would it make more sense to make this a meaningful value? e.g. `__no_item__` and use a constant?

Comment (lsmith77) [inline src/Jackalope/Transport/DoctrineDBAL/CachedClient.php]: I changed it to ``ItemNotFoundException``

File: src/Jackalope/Transport/DoctrineDBAL/CachedClient.php (hunk 1 above)
Comment (lsmith77) [inline src/Jackalope/Transport/DoctrineDBAL/CachedClient.php]: also fixed .. though I just copied this from the Client class where we currently do not use sprintf in most places. we should maybe go through it all in a separate PR

**Comment by dbu** [comment]
+1 to merge, makes sense to me.

**Comment by dbu** [comment]
should we also cache them in the jackrabbit client? hm, or should we have done this in ObjectManager instead of the client?

**Comment by lsmith77** [comment]
imho no .. or rather we should look into moving the entire caching into the ObjectManager .. not cache misses specifically. caching cache misses here is no problem here because we also have logic to clear the cache on write etc.

**Comment by dbu** [comment]
hm, good point. but do we have the same issue with jackrabbit client and should do something similar there?

**Comment by lsmith77** [comment]
well we do not have a client side cache for jackrabbit which is why yes, we should look at moving the caching into core of jackalope.
